dll/win32/browseui/CAutoComplete.cpp

```diff
@@ -25,664 +25,2124 @@
+// This message is sent when a window is being destroyed.
+LRESULT CACEditCtrl::OnDestroy(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL &bHandled)
+{
+    TRACE("CACEditCtrl::OnDestroy(%p)\n", this);
+    ATLASSERT(m_pDropDown);
+    CAutoComplete *pDropDown = m_pDropDown;
 
-            if (value)
-            {
-                *value = 0;
-                value++;
-                /* Now value contains the value and buffer the key */
-                res = RegOpenKeyExW(HKEY_CURRENT_USER, key, 0, KEY_READ, &hKey);
-        
-                if (res != ERROR_SUCCESS)
-                {
-                    /* if the key is not found, MSDN states we must seek in HKEY_LOCAL_MACHINE */
-                    res = RegOpenKeyExW(HKEY_LOCAL_MACHINE, key, 0, KEY_READ, &hKey);
-                }
-        
-                if (res == ERROR_SUCCESS)
-                {
-                    len = sizeof(result);
-                    res = RegQueryValueW(hKey, value, result, &len);
-                    if (res == ERROR_SUCCESS)
-                    {
-                        m_quickComplete = (WCHAR *)HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY, len * sizeof(WCHAR));
-                        wcscpy(m_quickComplete, result);
-                    }
-                    RegCloseKey(hKey);
-                }
-            }
+    // unhook word break procedure
+    HookWordBreakProc(FALSE);
 
-            HeapFree(GetProcessHeap(), 0, key);
-        }
-        else
-        {
-            TRACE("HeapAlloc Failed when trying to alloca %d bytes\n", (wcslen(pwzsRegKeyPath) + 1) * sizeof(WCHAR));
-            return S_FALSE;
-        }
-    }
+    // unsubclass because we don't watch any more
+    HWND hwndEdit = UnsubclassWindow();
 
-    if ((pwszQuickComplete) && (!m_quickComplete))
+    // close the drop-down window
+    if (pDropDown)
     {
-        m_quickComplete = (WCHAR *)HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY, (wcslen(pwszQuickComplete) + 1) * sizeof(WCHAR));
+        pDropDown->PostMessageW(WM_CLOSE, 0, 0);
+    }
 
-        if (m_quickComplete)
-        {
-            wcscpy(m_quickComplete, pwszQuickComplete);
-        }
-        else
+    return ::DefWindowProcW(hwndEdit, uMsg, wParam, lParam); // do default
+}
+
+// WM_GETDLGCODE
+// By responding to this message, an application can take control of a particular type of
+// input and process the input itself.
+LRESULT CACEditCtrl::OnGetDlgCode(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL &bHandled)
+{
+    TRACE("CACEditCtrl::OnGetDlgCode(%p)\n", this);
+    ATLASSERT(m_pDropDown);
+
+    LRESULT ret = DefWindowProcW(uMsg, wParam, lParam); // get default
+
+    if (m_pDropDown)
+    {
+        // some special keys need default processing. we handle them here
+        switch (wParam)
         {
-            TRACE("HeapAlloc Failed when trying to alloca %d bytes\n", (wcslen(pwszQuickComplete) + 1) * sizeof(WCHAR));
-            return S_FALSE;
+            case VK_RETURN: // [Enter] key
```

Comment:
I think everyone knows what these VK_xxx keys refer to.

Reply:
Deleted comments about `VK_*`.